### PR TITLE
filesystem/config: add support for configuring a biosboot partition (legacy BIOS or UEFI)

### PR DESCRIPTION
**Pull request by jouvin**
This PR provides the main part of the support for biosboot partitions (https://github.com/quattor/template-library-core/issues/72). It requires https://github.com/quattor/template-library-core/pull/126. The 2 PR altogether fix https://github.com/quattor/template-library-core/issues/72. In addition this PR brings UEFI support in the configuration description (also requires https://github.com/quattor/ncm-lib-blockdevices/issues/71 to be fixed for the actual support).

This PR contains also a some code refactoring/cleanup and several code fixes for long standing bugs uncovered when adding this new feature.

Heavily tested at LAL but tests by other sites are very welcome!

### CODE REVIEW

File: filesystem/config.pan [hunk 1]
```diff
@@ -5,8 +5,28 @@ unique template filesystem/config;
...
+ to redefine the default size of block devices defined in the default configuration
+values = template namespace
+default = FILESYSTEM_LAYOUT_CONFIG_SITE+'-init' or null if FILESYSTEM_LAYOUT_CONFIG_SITE is undefined
+required = no
+}
+variable FILESYSTEM_LAYOUT_CONFIG_INIT ?= if ( is_defined(FILESYSTEM_LAYOUT_CONFIG_SITE) ) {
+                                            if_exists(FILESYSTEM_LAYOUT_CONFIG_SITE + '-init');
+                                          } else {
+                                            null;
+                                          };
```
Comment (jrha) [inline filesystem/config.pan]: Can you get rid of all the left hand side white-space please.

File: filesystem/config.pan [hunk 2]
```diff
@@ -90,6 +122,106 @@ variable DISK_SWAP_SIZE ?= {
...
+desc =  default size for block device biosboot
+values = long
+default = 100 MB if using GPT label, 0 otherwise
+required = no
+}
+variable DISK_BIOSBOOT_BLOCKDEV_SIZE ?= if ( DISK_BOOT_ADD_BIOSBOOT_PART ) {
+                                          100*MB;
+                                        } else {
+                                          0;
+                                        };
```
Comment (jrha) [inline filesystem/config.pan]: Same here w.r.t white-space.

File: filesystem/config.pan [hunk 3]
```diff
@@ -99,22 +231,26 @@ variable DISK_SWAP_SIZE ?= {
 # designated by FILESYSTEM_LAYOUT_CONFIG_SITE (this variable is defined when this template is executed).
 # Key is an arbitrary name referenced by DISK_DEVICE_LIST.
 variable DISK_VOLUME_PARAMS ?= {
-  SELF['boot'] = nlist('size', 256*MB,
+  SELF['biosboot'] = nlist('size', DISK_BIOSBOOT_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 4]
```diff
@@ -99,22 +231,26 @@ variable DISK_SWAP_SIZE ?= {
...
+                       'type', 'partition',
+                       'flags', DISK_BIOSBOOT_PART_FLAGS,
+                       'device', DISK_BOOT_DEV+DISK_BOOT_PART_PREFIX+to_string(index('biosboot',DISK_BOOT_PARTS)+1));
+  SELF['boot'] = nlist('size', DISK_BOOT_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 5]
```diff
@@ -99,22 +231,26 @@ variable DISK_SWAP_SIZE ?= {
...
                        'mountpoint', '/boot',
                        'fstype', 'ext2',
                        'type', 'partition',
                        'device', DISK_BOOT_DEV+DISK_BOOT_PART_PREFIX+to_string(index('boot',DISK_BOOT_PARTS)+1));
-  SELF['home'] = nlist('size', 0*GB,
+  SELF['home'] = nlist('size', DISK_HOME_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 6]
```diff
@@ -99,22 +231,26 @@ variable DISK_SWAP_SIZE ?= {
...
                        'mountpoint', '/home',
                        'type', 'lvm',
-                       'volgroup', 'vg.01',
+                       'volgroup', DISK_VG01_VOLGROUP_NAME,
                        'device', 'homevol');
-  SELF['opt'] = nlist('size', 2*GB,
+  SELF['opt'] = nlist('size', DISK_OPT_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 7]
```diff
@@ -99,22 +231,26 @@ variable DISK_SWAP_SIZE ?= {
...
                       'mountpoint', '/opt',
                       'type', 'lvm',
-                      'volgroup', 'vg.01',
+                      'volgroup', DISK_VG01_VOLGROUP_NAME,
                       'device', 'optvol');
-  SELF['root'] = nlist('size', 1*GB,
+  SELF['root'] = nlist('size', DISK_ROOT_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 8]
```diff
@@ -123,29 +259,29 @@ variable DISK_VOLUME_PARAMS ?= {
                        'fstype', 'swap',
                        'type', 'partition',
                        'device', DISK_BOOT_DEV+DISK_BOOT_PART_PREFIX+to_string(index('swap',DISK_BOOT_PARTS)+1));
-  SELF['swareas'] = nlist('size', 0*GB,
+  SELF['swareas'] = nlist('size', DISK_SWAREAS_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 9]
```diff
@@ -123,29 +259,29 @@ variable DISK_VOLUME_PARAMS ?= {
...
                           'mountpoint', '/swareas',
                           'type', 'lvm',
-                          'volgroup', 'vg.01',
+                          'volgroup', DISK_VG01_VOLGROUP_NAME,
                           'device', 'swareasvol');
-  SELF['tmp'] = nlist('size', 1*GB,
+  SELF['tmp'] = nlist('size', DISK_TMP_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 10]
```diff
@@ -123,29 +259,29 @@ variable DISK_VOLUME_PARAMS ?= {
...
                       'mountpoint', '/tmp',
                       'type', 'lvm',
-                      'volgroup', 'vg.01',
+                      'volgroup', DISK_VG01_VOLGROUP_NAME,
                       'device', 'tmpvol');
-  SELF['usr'] = nlist('size', 5*GB,
+  SELF['usr'] = nlist('size', DISK_USR_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 11]
```diff
@@ -123,29 +259,29 @@ variable DISK_VOLUME_PARAMS ?= {
...
                       'mountpoint', '/usr',
                       'type', 'lvm',
-                      'volgroup', 'vg.01',
+                      'volgroup', DISK_VG01_VOLGROUP_NAME,
                       'device', 'usrvol');
-  SELF['var'] = nlist('size', -1,
+  SELF['var'] = nlist('size', DISK_VAR_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 12]
```diff
@@ -123,29 +259,29 @@ variable DISK_VOLUME_PARAMS ?= {
...
                       'mountpoint', '/var',
                       'type', 'lvm',
-                      'volgroup', 'vg.01',
+                      'volgroup', DISK_VG01_VOLGROUP_NAME,
                       'device', 'varvol');
-  SELF['vg.01'] = nlist('size', -1,
-                        'type', 'vg',
-                        'devices', list(DISK_BOOT_DEV+DISK_BOOT_PART_PREFIX+to_string(index('lvm',DISK_BOOT_PARTS)+1)));
+  SELF[DISK_VG01_VOLGROUP_NAME] = nlist('size', DISK_VG01_BLOCKDEV_SIZE,
```
Comment (jrha) [inline filesystem/config.pan]: `dict(`

File: filesystem/config.pan [hunk 13]
```diff
@@ -339,18 +475,28 @@ variable DISK_PART_BY_DEV = {
...
         if ( !exists(SELF['partitions'][phys_dev]) ) {
-          # Use 2 separate nlist, size and part_num, to ease processing later (partitions_add() requires a nlist of
-          # partitions where the key is the partition name and the value the size). In each list,
-          # the key is the partition name.
-          SELF['partitions'][phys_dev] = nlist('size', nlist(),
+          # Build 2 separate nlist, part_list and part_num, the key being the partition name in each
+          # list. part_list will be passed to partitions_add() which requires a dict of
+          # partitions where the key is a partitionname and the value the partition parameters (as a dict).
+          # part_num is a transient dict used internally to do the partition final numbering.
+          SELF['partitions'][phys_dev] = nlist('part_list', nlist(),
```
Comment (jrha) [inline filesystem/config.pan]: `dict(` `dict()`

File: filesystem/config.pan [hunk 14]
```diff
@@ -339,18 +475,28 @@ variable DISK_PART_BY_DEV = {
...
                                                'part_num', nlist(),
```
Comment (jrha) [inline filesystem/config.pan]: `dict()`

File: filesystem/config.pan.new [hunk 15]
```diff
@@ -0,0 +1,804 @@
+# Template allow to configure partitionning with quite a lot of flexibility,
```
Comment (jrha) [inline filesystem/config.pan.new]: partition~~n~~ing

So which template do you recommend using? Why not just replace `filesystem/config.pan`?

Whole template could do with `nlist` → `dict` and removal of `include { }`.

**Comment by jouvin** [comment]
@jrha Why this PR has been postponed to 16.12 (except that I was may be unresponsive!)... This is an important one as it is a blocking issue for installing machines with disks larger than 2 TB. I'll try to address your comments today.

File: filesystem/config.pan (hunk 1 above)
Comment (jouvin) [inline filesystem/config.pan]: I don't really understand your point... This is the standard formatting used through the template library when the `if` follows the equal sign...

File: filesystem/config.pan (hunk 2 above)
Comment (jouvin) [inline filesystem/config.pan]: idem

File: filesystem/config.pan.new (hunk 15 above)
Comment (jouvin) [inline filesystem/config.pan.new]: I don't understand your question... There used to be a documentation about this template on the old wiki that may be worth migrating/rewriting... The basic idea is that you never have to replace `filesystem/config` that does a lot of things for you. You just need to define another template (called layout) where you define a few variables to steer `filesystem/config`.

**Comment by stdweird** [comment]
@jouvin not to delay this any further, but maybe it's an easy fix/addition: UEFI requires a uefi partition, similar to the biosboot for legacy bios with gpt: a small vfat partition with boot flag set. UEFI does not require a biosboot partiton, so maybe a simple boolean to convert the bioboot partiton into a uefi one might be intersting.

if this is too much to change in this PR, do it later. but as it is very similar, it might be easy to add.

**Comment by jouvin** [comment]
@stdweird I'll try to add it but will probably not be able to test it.

File: filesystem/config.pan (hunk 1 above)
Comment (jrha) [inline filesystem/config.pan]: That may be, but it's horrible, it makes all the lines far too long and makes them dependent on the length of the left hand side of the assignment, which when it changes causes the following lines to look like they have code changes.

To me, this is both easier to read and maintain.

```pan
variable FILESYSTEM_LAYOUT_CONFIG_INIT ?= if (is_defined(FILESYSTEM_LAYOUT_CONFIG_SITE)) {
    if_exists(FILESYSTEM_LAYOUT_CONFIG_SITE + '-init');
} else {
    null;
}; 
```

But I can't be bothered to argue the point any more and I certainly won't delay the PR for it.

**Comment by jrha** [comment]
@jouvin the PR was delayed precisely because it didn't seem like you were going to be able to update it in time, indeed if I was able to stick to the schedule these changes would have been too late.

**Comment by jouvin** [comment]
@stdweird UEFI support should be pretty easy to add according to your description. Please, could you tell me:
- is the partition name imposed? If yes what should it be? (partition name is not imposed for GPT)
- are there some partition flags to set? Which ones?
- what is the recommended partition size? (we use 100 MB for GPT biosboot)
- Should the file system format be specified or will the partition be formatted by UEFI boot ?

**Comment by stdweird** [comment]
* UEFI requires gpt label, so any name is fine (maybe efi or uefi)
* the boot flag has to be set
* recommended size is 200MB (for biosboot, 1MB is fine, 100MB is wasteful)
* i don't know if it's formatted by the tool, should be FAT32 and mounted under `/boot/efi` (i don't know if `ncm-lib-blockdevices` has fat32 support for `%pre`)

File: filesystem/config.pan [hunk 16]
```diff
@@ -151,13 +151,29 @@ variable DISK_SWAP_SIZE ?= {
...
 values = long
-default = 100 MB
+default = 100 MB for legacy BIOS, 200 MB for UEFI
 required = no
 }
-variable DISK_BIOSBOOT_BLOCKDEV_SIZE_DEFAULT ?= 100*MB;
+variable DISK_BIOSBOOT_BLOCKDEV_SIZE_DEFAULT ?= if ( DISK_BIOS_TYPE_UEFI ) {
+                                       200*MB;
+                                     } else {
+                                       100*MB;
```
Comment (jouvin) [inline filesystem/config.pan]: This is what I already heard several times but Adrien who gave me his suggestions told me 100 MB... I'll check with him and update it.

**Comment by jouvin** [comment]
Commit https://github.com/quattor/template-library-standard/pull/85/commits/541c21d34f2ba38b8ef4818f19acc8827c16f274 brings UEFI support in the profile but requires  https://github.com/quattor/ncm-lib-blockdevices/issues/71 to be fixed to be really effective.

**Comment by jouvin** [comment]
Would be really good to have it in 16.10 if somebody can review it...

**Comment by jrha** [comment]
Looks reasonable enough to me, but we don't use these templates for file-system configuration.

**Comment by jouvin** [comment]
I think that they were used in Bruxelles or UGent... But they have been heavily tested, so if nobody is using them and is able to review them, I'd suggest to merge the PR...

**Pull request merged**